### PR TITLE
Improve c++17 compile

### PR DESCRIPTION
**Pull request by wduminy**
Follows on from #183

Fixes use of non-standard shuffle.

Changed the build to use `c++17`

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,6 @@
 build
 .kdev4
+.vscode
```
Comment (AlexanderFabisch) [inline .gitignore]: Is this necessary?

File: CMakeLists.txt [hunk 2]
```diff
@@ -51,6 +52,10 @@ set(OPENANN_BRIEF_DESCRIPTION "An open source library for artificial neural netw
 current_time(OPENANN_TIME)
 set(OPENANN_OPTIMIZATION_FLAGS)
 set(OPENANN_PYTHON_COMPILE_ARGS)
+if(CPP_17)
+	compiler_check_add_flag("-std=c++17")
+	compiler_check_add_flag("-Dregister=") # a work-around to get rid of the register keyword
```
Comment (AlexanderFabisch) [inline CMakeLists.txt]: I think we can delete register from the whole codebase. It is not really required I guess.

File: OpenANN/util/Random.h [hunk 3]
```diff
@@ -110,7 +119,7 @@ class RandomNumberGenerator
 #if __cplusplus < 201300L		
     std::random_shuffle(result.begin(), result.end());
 #else
-    std::shuffle(result.begin(), result.end());
+    std::shuffle(result.begin(), result.end(),m_generator);
```
Comment (AlexanderFabisch) [inline OpenANN/util/Random.h]: Please add a space here.

**Review by AlexanderFabisch — CHANGES_REQUESTED** [review]
The string output of Eigen matrices seems to have changed in a recent version that is the reason why one of the test fails. Could you fix that?

[IODataSetTestCase.cpp:98](https://github.com/OpenANN/OpenANN/blob/master/test/IODataSetTestCase.cpp#L98) - should be `1.5   0` with 3 spaces between 1.5 and 0

**Comment by AlexanderFabisch** [comment]
We should actually have CI for C++17 but I guess that would require setting up a docker image for CircleCI which is work for future me.

File: .gitignore (hunk 1 above)
Comment (wduminy) [reply .gitignore]: Minor convenience - I do not use kdev4, but I use vs code.

File: CMakeLists.txt (hunk 2 above)
Comment (wduminy) [reply CMakeLists.txt]: Unfortunately `lib/CMA-ESpp/cma-es/utils.h` uses `register` in the header, so the work-around seems to be needed.

**Comment by wduminy** [comment]
@AlexanderFabisch I have given this PR a little more thought.  

After this change, client code would not be able switch on `c++17` on their code without specifically compiling the library for c++17.  

In fact, it will probably crash when you compile the library without c++17 and use it with c++17.

It is tricky to fix this problem because `generateIndices` is a template function.  It will probably cause a breaking change.

**Comment by wduminy** [comment]
@AlexanderFabisch I guessed you would agree with the above comment; so I now OpenANN depends  on c++17.  This was a good idea.  Some of the unit tests fail.  Probably a good thing.

**Comment by wduminy** [comment]
Decided to use a "mixed approach" on Random.cpp.  Now unit tests are all fine :-)

File: src/DataSetView.cpp [hunk 4]
```diff
@@ -106,11 +103,7 @@ void split(std::vector<DataSetView>& groups, DataSet& dataset, double ratio,
 
   if(shuffling)
 	{
-		#if __cplusplus < 201300L		
-			std::random_shuffle(indices.begin(), indices.end());
-		#else
-			std::shuffle(indices.begin(), indices.end());
-		#endif	
+    std::shuffle(indices.begin(), indices.end(), randomGenerator);
```
Comment (AlexanderFabisch) [inline src/DataSetView.cpp]: indentation does not match

**Comment by AlexanderFabisch** [comment]
Apart from the last comment everything seems to be ready for merge.

**Review by AlexanderFabisch — APPROVED** [review]

**Comment by AlexanderFabisch** [comment]
Thanks @wduminy

**Pull request merged**